src/libsyntax/parse/parser.rs

```diff
@@ -8712,22 +8713,43 @@ impl<'a> Parser<'a> {
     ///
     /// The arguments of the function are replaced in HIR lowering with the arguments created by
     /// this function and the statements created here are inserted at the top of the closure body.
-    fn construct_async_arguments(&mut self, asyncness: &mut Spanned<IsAsync>, decl: &FnDecl) {
+    fn construct_async_arguments(&mut self, asyncness: &mut Spanned<IsAsync>, decl: &mut FnDecl) {
```

Comment:
@davidtwco This is the hack that we'd want to move to HIR lowering but we can't because of `rustc_resolve` interacting with upvars, right?

Reply:
@eddyb Yeah, we'll presumably want to lower to the same HIR in the end. Because of the interaction with `rustc_resolve`, it's easier to insert the statements/replace the arguments in the AST during name resolution so it can make the correct locals resolve to `Def::Upvar`. If we didn't construct them here, then these same types would need to be constructed in `rustc_resolve` and in HIR lowering. 

Other places, like the def collector, that use these constructed types would be able to get away with just using `NodeId`s that we'd create here (much like the `IsAsync` type worked previously), and we'd only construct the actual statements and arguments during HIR lowering.